source/uri-options/tests/connection-pool-options.yml

```diff
@@ -46,3 +47,21 @@ tests:
       auth: ~
       options:
           minPoolSize: 0 
+
+    -
+      description: "maxConnecting=0 causes an error"
+      uri: "mongodb://example.com/?maxConnecting=0"
+      valid: false
+      warning: false
```

Comment:
For both of these tests, this should be:
```
      valid: true
      warning: true
```

And the test names should change from "causes an error" to "causes a warning". See "Too low maxIdleTimeMS causes a warning" for example.

Reply:
both of these tests failed in c# as well because we don't have a defined rule whether we should make the whole possible value validation on the connectionString level or no, sometimes we do it, sometimes no. To fix it, I added a value validation for this particular case

Reply:
What is the reasoning behind not throwing an exception (or erring out) when `maxConnecting` is an invalid value? The .NET/C# driver doesn't have a mechanism to log warnings at the moment.

Reply:
I'm not sure why it's like this but that's how Python works and it's how we test the other options:
```python
>>> c = MongoClient("mongodb://example.com/?maxConnecting=-1")
/Users/shane/git/mongo-python-driver/pymongo/common.py:771: UserWarning: The value of maxconnecting must be a positive integer
  warnings.warn(str(exc))
```

Reply:
@ShaneHarvey I also do not understand the problem with reacting to an invalid value of `maxConnecting` with an error. We do so for `srvMaxHosts`, for example. I do see that for `maxPoolSize` we produce a warning instead, but I don't see anything in https://github.com/mongodb/specifications/blob/master/source/uri-options/uri-options.rst that precludes producing an error and explains why a warning must be produced instead. I also fail to think of a good reason to produce a warning instead of an error.

Reply:
srvMaxHosts also uses warning:true, valid:true like I'm suggesting here:
[```
    {
      "description": "SRV URI with negative integer for srvMaxHosts",
      "uri": "mongodb+srv://test1.test.build.10gen.cc/?srvMaxHosts=-1",
      "valid": true,
      "warning": true,
      "hosts": null,
      "auth": null,
      "options": {}
    },
    {
      "description": "SRV URI with invalid type for srvMaxHosts",
      "uri": "mongodb+srv://test1.test.build.10gen.cc/?srvMaxHosts=foo",
      "valid": true,
      "warning": true,
      "hosts": null,
      "auth": null,
      "options": {}
    },
```](https://github.com/mongodb/specifications/blob/84bdb3f59892944efff7ad2b13d6fb1278dd81ba/source/uri-options/tests/srv-options.yml#L25-L38)

Reply:
I think it odd that we don't err on bad URI values in certain cases, but that's a bigger discussion not specific to `maxConnecting`. Let's go with a warning for invalid values of `maxConnecting` for consistency with other options.

Reply:
And here it does not :) https://github.com/mongodb/specifications/blob/84bdb3f59892944efff7ad2b13d6fb1278dd81ba/source/uri-options/tests/srv-options.yml#L39-L42

I think we need a better argument in favour of not producing an error than the Python driver implementation detail.
There are both warnings and errors in the tests, and I think that errors are more explicit and produce less surprising results for a user than warnings (the driver does not end up configured differently from how a user expected it to be). This is why I think producing an error is preferable.

Reply:
That case is different though. In that case it's the schema+srvMaxHosts that's an error. I think it's outside the scope of this ticket to change the behavior here. If you'd like to change the warning/error behavior we should create a new drivers ticket and address the problem consistently for all options. But connection string spec requires that invalid values for uri options will raise a warning, not an error:
```
Any invalid Values for a given key MUST be ignored and MUST log a WARN level message. For example:

Unsupported value for "fsync" : "ifPossible"
```
https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.rst#values

Reply:
>What is the reasoning behind not throwing an exception

the reason that we almost never does it for just `number` values on this level, so consistency. 

>The .NET/C# driver doesn't have a mechanism to log warnings at the moment.

we consider warnings as errors if the exception is triggered and ignore it if the exception has not been raised.

Reply:
Ok, @jyemin pointed out that the Connection String Spec says [_"Any invalid Values for a given key MUST be ignored and MUST log a WARN level message"_](https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.rst#:~:text=Any%20invalid%20Values%20for%20a%20given%20key%20MUST%20be%20ignored%20and%20MUST%20log%20a%20WARN%20level%20message). It does not explain the reasoning behind the requirement, but explains the reasoning behind a somewhat related requirement: [_"Q: Why WARN level warning on unknown options rather than throwing an exception?"_](https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.rst#:~:text=Q%3A%20Why%20WARN%20level%20warning%20on%20unknown%20options%20rather%20than%20throwing%20an%20exception%3F).

Despite there being no known explanation for the requirement of not producing errors for invalid values, we have an explicit requirement in the spec, which resolves the argument. I will change the test such that it expects a warning.

Reply:
@ShaneHarvey I have just seen your [reply](https://github.com/mongodb/specifications/pull/1098#discussion_r749675228). Yes, that part of the spec is exactly the reason we have to produce a warning.

Reply:
I updated the test. While doing so, I discovered the following:
1. The Java driver throws errors for invalid values of all options and ignores all spec tests that contain `"warning": true`.
2. The [URI-options test runner spec](https://github.com/mongodb/specifications/blob/master/source/uri-options/tests/README.rst) & tests are confusing regarding what it means for a connection string to be "valid". For example, a string that produces a warning during validation is invalid according to [_"Testing whether a URI is valid or not requires testing whether URI parsing (or MongoClient construction) causes a warning due to a URI option being invalid"_](https://github.com/mongodb/specifications/blob/master/source/uri-options/tests/README.rst#:~:text=Testing%20whether%20a%20URI%20is%20valid%20or%20not%20requires%20testing%20whether%20URI%20parsing%20(or%20MongoClient%20construction)%20causes%20a%20warning%20due%20to%20a%20URI%20option%20being%20invalid), yet many tests use `valid: true warning: true`.

Reply:
> Testing whether a URI is valid or not requires testing whether URI parsing (or MongoClient construction) causes a warning due to a URI option being invalid

This line is incorrect and we should file a new ticket to change it.